### PR TITLE
Recaptcha function

### PR DESCRIPTION
**Pull request by PriyaRajarathinam**
Problem
=======

As an engineer, I want to verify that people who log in or sign up are real users and not bots.

https://app.asana.com/0/1197738625416321/1197738625416343

Solution
========

Added an endpoint that takes in the token generated by Google's Recaptcha and validates it with Google's Recaptcha server-side verification API

Change summary:
---------------
* Added the siteVerify function 

Steps to Verify:
----------------
1. npm run serve 
2. Generate a token by creating a button integrated with the client-side key
3. Call the endpoint by adding token as the key name  and  its value as a key-value pair
4. It will return a JSON body which indicates whether the authentication was a success or not

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -44,6 +44,6 @@ jobs:
       - name: Deploy to Firebase
         uses: w9jds/firebase-action@master
         with:
-          args: deploy --only functions #TODO: specify function you want to deploy
+          args: deploy --only functions:siteVerify #TODO: specify function you want to deploy
```
Comment (w-gao) [inline .github/workflows/main.yaml]: Is this still a TODO?

File: functions/index.js [hunk 2]
```diff
@@ -1,9 +1,64 @@
...
+    
+    const [accessResponse] = await client.accessSecretVersion({
+        name: "projects/cruzhacks-4a899/secrets/RecaptchaSecretKey/versions/latest",
+    });
+    const secret = accessResponse.payload.data.toString('utf8');
+    const siteVerifyUrl = "https://www.google.com/recaptcha/api/siteverify";
+    
+    // Check if token is in headers 
+    if (!(Object.keys(request.headers).includes("token"))) {
+        return response.status(400).send("Token is missing");
```
Comment (w-gao) [inline functions/index.js]: Should this be wrapped in JSON so our response format is consistent?

File: functions/index.js [hunk 3]
```diff
@@ -1,9 +1,64 @@
...
+        // Handle response from Google's recaptcha API
+        if (json['error-codes']) {
+            if ('invalid-input-response' in json['error-codes'] && 'invalid-input-secret' in json['error-codes']){
+                return response.status(500).send('Invalid token provided and incorrect secret')
+            }
+            else if ('invalid-input-response' in json['error-codes']){
+                return response.status(400).send('Invalid token provided')
+            }
+            else if ('invalid-input-secret' in json['error-codes']){
+                return response.status(500).send('incorrect secret')
```
Comment (w-gao) [inline functions/index.js]: Same as above

File: functions/index.js [hunk 4]
```diff
@@ -1,9 +1,64 @@
...
+            }
+        }  
+        return response.status(200).send(json)
+    })
+    .catch(error => {
+        console.log(error)
+        return response.status(500).send(error);
+    }) 
+})
```
Comment (w-gao) [inline functions/index.js]: EOF

**Review by w-gao — COMMENTED** [review]
This looks like it would work. Just a few comments. Great work overall!

File: functions/tests/index.test.js [hunk 5]
```diff
@@ -1,4 +1,154 @@
...
+      },
+      send: ({ error, status, message }) => {
+        chai.assert.equal(error, true);
+        chai.assert.equal(status, 400);
+        chai.assert.equal(message, "Request Timed Out or Sent Duplicate Key");
+        done();
+      },
+    };
+    siteVerify.siteVerify(req, res);
+  });
```
Comment (ilopezro) [inline functions/tests/index.test.js:129]: Added this test case!

**Review by ilopezro — APPROVED** [review]
This looks really good! 🚢

**Pull request merged**